File: BOJ/python3/1743.py
'''
제목
'''
import sys
import time
start = time.time()  # 시작 시간
sys.stdin = open("input.txt", "r")  # input.txt 파일로 입력


def dfs(pos):
    global cnt, size

    if cnt > size:
        size = cnt

    r, c = pos
    for i in range(4):
        next_r = r + dr[i]
        next_c = c + dc[i]
        if 0 < next_r <= N and 0 < next_c <= M:
            if mat[next_r][next_c] == '#' and visit[next_r][next_c] == 0:
                visit[next_r][next_c] = 1
                cnt += 1
                dfs((next_r, next_c))


N, M, K = map(int, input().split())
mat = [['0']*(M+1)]
for n in range(N):
    row = ['0'] + ['.']*M
    mat.append(row)
for _ in range(K):
    r, c = map(int, input().split())
    mat[r][c] = '#'

ans = 0
cnt, size = 0, 0
dr = [-1, 0, 1, 0]
dc = [0, 1, 0, -1]
visit = [[0]*(M+1) for _ in range(N+1)]
for r in range(1, N+1):
    for c in range(1, M+1):
        if mat[r][c] == '#' and visit[r][c] == 0:
            visit[r][c] = 1
            cnt = 1
            dfs((r, c))

            if size > ans:
                ans = size

            cnt, size = 0, 0

print(ans)
# 현재시각 - 시작시간 = 실행 시간(sec)
print("--------------------")
print(f"Elapsed Time: {round(time.time() - start, 3)}s")

File: BOJ/nodejs/1743.js
/*
음식물 피하기
boj.kr/1743
 */
const start = new Date().getTime();

const fs = require("fs");
const input = fs.readFileSync("input.txt").toString().split("\n");
//const input = fs.readFileSync("/dev/stdin").toString().split("\n");

const [N, M, K] = input[0].split(" ").map(Number); // [세로, 가로, 쓰레기 개수]
const mat = Array.from({ length: N + 1 }, () => Array(M + 1).fill("."));
for (let k = 0; k < K; k++) {
  let [r, c] = input[1 + k].split(" ").map(Number);
  mat[r][c] = "#";
}

let ans = 0;
let size;
const visited = Array.from({ length: N + 1 }, () => Array(M + 1).fill(0));
const dr = [-1, 0, 1, 0];
const dc = [0, 1, 0, -1];
const dfs = (pos) => {
  let [r, c] = [pos[0], pos[1]];
  visited[r][c] = 1;
  size++;
  ans = Math.max(ans, size);

  for (let i = 0; i < 4; i++) {
    let [nr, nc] = [pos[0] + dr[i], pos[1] + dc[i]];
    if (nr >= 1 && nr <= N && nc >= 1 && nc <= M)
      if (mat[nr][nc] === "#" && visited[nr][nc] === 0) dfs([nr, nc]);
  }
};

for (let r = 1; r < N + 1; r++) {
  for (let c = 1; c < M + 1; c++) {
    if (mat[r][c] === "#" && visited[r][c] === 0) {
      size = 0;
      dfs([r, c]);
    }
  }
}

console.log(ans);

const end = new Date().getTime();
console.log("--------------------");
console.log(`Elapsed Time: ${(new Date().getTime() - start) / 1000}s`);
